js.go

```diff
@@ -806,20 +812,23 @@ func (js *js) handleAsyncReply(m *Msg) {
 		js.mu.Unlock()
 		return
 	}
-	// Remove
-	delete(js.pafs, id)
 
-	// Check on anyone stalled and waiting.
-	if js.stc != nil && len(js.pafs) < js.opts.maxpa {
-		close(js.stc)
-		js.stc = nil
+	closeStc := func() {
+		// Check on anyone stalled and waiting.
+		if js.stc != nil && len(js.pafs) < js.opts.maxpa {
+			close(js.stc)
+			js.stc = nil
+		}
 	}
-	// Check on anyone one waiting on done status.
-	if js.dch != nil && len(js.pafs) == 0 {
-		dch := js.dch
-		js.dch = nil
-		// Defer here so error is processed and can be checked.
-		defer close(dch)
+
+	closeDch := func() {
+		// Check on anyone one waiting on done status.
+		if js.dch != nil && len(js.pafs) == 0 {
+			dch := js.dch
+			js.dch = nil
+			// Defer here so error is processed and can be checked.
+			defer close(dch)
```

Comment:
This defer does not work as expected since it'll be executed at the end of `closeDch`, not the caller. While it would be good to extract this func since it's repeated twice, I do think that we should make sure that `close(dch)` is executed at the very end to first send signals on the `pubAckFuture` and only later to `PublishAsyncComplete()`